spock-core/src/main/java/org/spockframework/runtime/ParameterizedSpecRunner.java

```diff
@@ -132,20 +136,45 @@ private int estimateNumIterations(Object[] dataProviders) {
     return result == Integer.MAX_VALUE ? -1 : result;
   }
 
-  private void runIterations(Iterator[] iterators, int estimatedNumIterations) {
-    if (runStatus != OK) return;
+  private void runIterations(final FeatureInfo currentFeature, final Iterator[] iterators, final int estimatedNumIterations) {
+    if (runStatus.get() != OK) return;
+
+    boolean atLeastOneIteration = false;
+
+    Scheduler schedulerForRunIterations = this.scheduler.deriveScheduler(!currentFeature.isReportIterations() || !currentFeature.isSupportParallelExecution());
+
+    while (haveNext(currentFeature, iterators)) {
+      atLeastOneIteration = true;
+      final Object[] dataValues = nextArgs(currentFeature, iterators);
 
-    while (haveNext(iterators)) {
-      initializeAndRunIteration(nextArgs(iterators), estimatedNumIterations);
+      if (haveNext(currentFeature, iterators)) {
+        schedulerForRunIterations.schedule(new Runnable() {
+          @Override
+          public void run() {
+            runStatus.set(OK);
```

Comment:
> If scheduler is thread-pool scheduler - we should set(OK) unconditionally (bad status can come from another test)
That is the problem I'm talking about, you overwrite the status of another test, maybe before it was evaluated, either we don't really need `runStatus` or we have to make sure it is the correct one.